<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>DIAlign: DIAlign: Alignment of MS/MS chromatograms</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    extensions: ["tex2jax.js"],
    jax: ["input/TeX","output/HTML-CSS"],
});
</script><script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML/MathJax.js"></script>
>>>>>>> 303e7345e09e542cd07ff5f8aba17153f6ad470d
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">DIAlign
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">DIAlign: Alignment of MS/MS chromatograms </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#section0">MS/MS Chromatogram group</a></li>
<li class="level1"><a href="#section1">Similarity Matrix</a><ul><li class="level2"><a href="#subsection1">Gap Penalty</a></li>
<li class="level2"><a href="#subsection2">Constraint</a></li>
</ul>
</li>
<li class="level1"><a href="#section2">Affine alignment</a><ul><li class="level2"><a href="#subsecce">Initialization</a></li>
<li class="level2"><a href="#subsection3">Best score calculation</a></li>
<li class="level2"><a href="#subsection4">Traceback</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p><a class="el" href="Initialization.html">Initialization</a> <a class="el" href="Matrices.html">Matrices</a></p>
<p>&lt;script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"&gt;&lt;/script&gt;</p>
<p>Data Independent Acqusition (DIA) is a widely used method in mass-spectrometry based proteomics. <a class="el" href="namespaceDIAlign.html" title="Generic namespace for all classes and functions of DIAlign. ">DIAlign</a> is a software for the alignment of MS/MS chromatograms across runs. We have <a class="el" href="namespaceDIAlign.html" title="Generic namespace for all classes and functions of DIAlign. ">DIAlign</a> namespace that includes functions for building similarity matrix from raw MS2 chromatograms. On the similarity matrix, affine-alignment is performed using dynamic programming. Aligned indices, alignment path through the matrix and intermediate cumulative scoring matrices are included in the alignment object.</p>
<h1><a class="anchor" id="section0"></a>
MS/MS Chromatogram group</h1>
<p>A chromatogram is a time-series signal with time and intensity. We define a group as a collection of chromatograms having the same time values. The algorithm aligns a chromatogram group from run A to the same of run B. Both groups must contain the same number of chromatograms.</p>
<h1><a class="anchor" id="section1"></a>
Similarity Matrix</h1>
<p>The first step is to build a similatry score matrix between both chromatogram groups. Function <a class="el" href="namespaceDIAlign_1_1SimilarityMatrix.html#a3be1fab9e297669efbe44bac2a402be9">getSimilarityMatrix </a> returns the score matrix. Following similarity score methods are implemented: <em>dotProductMasked</em>, <em>dotProduct</em>, <em>cosineAngle</em>, <em>cosine2Angle</em>, <em>euclideanDist</em>, <em>covariance</em>, <em>correlation</em>.</p>
<h2><a class="anchor" id="subsection1"></a>
Gap Penalty</h2>
<p>Gap penalty is calculated from the distribution of similarity scores. <a class="el" href="namespaceDIAlign.html#ad5de813b6d7cb93540d1716509666523">getGapPenalty </a> function calculates the penalty score as a certain quantile value from the similarity score distribution.</p>
<h2><a class="anchor" id="subsection2"></a>
Constraint</h2>
<p>The hybrid alignment requires constraing the similarity matrix to avoid a large drift. To achieve this a constraining function is calculated externally and scores outside the certain distance from this function is penalized. <a class="el" href="namespaceDIAlign_1_1ConstrainMatrix.html">ConstrainMatrix </a> has functions to implement it. In the given example below, scoring matrix is represented as a grid between (A1, B1) and (A2, B2). <code>X</code> represents unconstrained scores, whereas, whitespace region is constrained.</p>
<pre class="fragment">    B1              B2
  A1 +---------------+
     |XX             |
     | XXX           |
     |  XXX          |
     |   XXX         |
     |    XXX        |
     |     XXX       |
     |      XXX      |
     |       XXX     |
  A2 +---------------+
</pre><h1><a class="anchor" id="section2"></a>
Affine alignment</h1>
<p>We use dynamic programming for finding highest-scoring path through the score matrix. To avoid execursions in horizontal or vertical direction, affine-alignment is used with low gap-opening penalty and high gap-extension penalties.</p>
<p>Given a similarity score matrix s of size n x m, or affine-alignment we calculate three matrices M, A and B each of size (n+1) x (m+1).</p><ul>
<li>Matrix M stores the highest score for the alignment of A<sub>i</sub> to B<sub>j</sub>.</li>
<li>Matrix A stores the best score for the alignment of A<sub>i</sub> to B<sub>j</sub> given A<sub>i</sub> is aligned to a gap.</li>
<li>Matrix B stores the highest score for the alignment of B<sub>j</sub> to A<sub>i</sub>, given B<sub>j</sub> is aligned to a gap.</li>
</ul>
<p>Example: We have following similarity matrix. Gap opening penalty = 22, gap extension penalty = 7 </p><table class="doxtable">
<tr>
<th>s </th><th>1 </th><th>2 </th><th>3 </th><th>4 </th><th>5  </th></tr>
<tr>
<td>1 </td><td>10 </td><td>-2 </td><td>-2 </td><td>-2 </td><td>-2 </td></tr>
<tr>
<td>2 </td><td>-2 </td><td>10 </td><td>-2 </td><td>-2 </td><td>-2 </td></tr>
<tr>
<td>3 </td><td>-2 </td><td>-2 </td><td>-2 </td><td>10 </td><td>-2 </td></tr>
</table>
<h2><a class="anchor" id="subsecce"></a>
Initialization</h2>
<p>The corresponding M, A and B matrices are initialized with some set of rules for overlap and non-overlap alignment. See <a class="el" href="Initialization.html">Initialization</a> for an example.</p>
<h2><a class="anchor" id="subsection3"></a>
Best score calculation</h2>
<p>Highest score for each cell is calculated using the following equations: \( M_{i,j} = max \left\{\begin{matrix} M_{i-1,j-1} + s_{i,j} \\ A_{i-1,j-1} + s_{i,j} \\ B_{i-1,j-1} + s_{i,j} \end{matrix}\right\} \)</p>
<p>\( A_{i,j} = max \left\{\begin{matrix} M_{i-1,j} - gapOpen \\ A_{i-1,j} - gapExtension \\ B_{i-1,j} - gapOpen \end{matrix}\right\} \)</p>
<p>\( B_{i,j} = max \left\{\begin{matrix} M_{i,j-1} - gapOpen \\ A_{i,j-1} - gapOpen \\ B_{i,j-1} - gapExtension \end{matrix}\right\} \)</p>
<p>It is clear that these matrices restricts the direction. Matrix M only allows diagonal path (upperleft to downright), matrix A allows up to down and matrix B permits left to right path only. The example matrices are presented here: <a class="el" href="Matrices.html">Matrices</a></p>
<h2><a class="anchor" id="subsection4"></a>
Traceback</h2>
<p>After calculating best cumulative score, we perform traceback and get the alignment path.</p>
<p>Alignment paths for Overlap alignment:</p>
<div class="image">
<img src="overlap.svg" alt="overlap.svg"/>
<div class="caption">
Overlap Alignment</div></div>
<p> Alignment paths for non-Overlap alignment:</p>
<div class="image">
<img src="nonoverlap.svg" alt="nonoverlap.svg"/>
<div class="caption">
Non-overlap Alignment</div></div>
 <dl class="section see"><dt>See also</dt><dd><ul>
<li>Gupta S, Ahadi S, Zhou W, Röst H. "DIAlignR Provides Precise Retention Time Alignment Across Distant Runs in DIA and Targeted Proteomics." Mol Cell Proteomics. 2019 Apr;18(4):806-817. doi: <a href="https://doi.org/10.1074/mcp.TIR118.001132">https://doi.org/10.1074/mcp.TIR118.001132</a> Epub 2019 Jan 31.</li>
<li>CNPN 2018 Poster doi: <a href="https://doi.org/10.6084/m9.figshare.6200837.v1">https://doi.org/10.6084/m9.figshare.6200837.v1</a></li>
<li>HUPO 2018 Poster doi: <a href="https://doi.org/10.6084/m9.figshare.7121696.v2">https://doi.org/10.6084/m9.figshare.7121696.v2</a></li>
<li><a href="https://www.coursera.org/lecture/bioinformatics-pku/alignment-with-affine-gap-penalty-and-calculation-of-time-complexity-of-the-0ya7X">https://www.coursera.org/lecture/bioinformatics-pku/alignment-with-affine-gap-penalty-and-calculation-of-time-complexity-of-the-0ya7X</a></li>
<li><a href="https://www.cs.cmu.edu/~ckingsf/bioinfo-lectures/gaps.pdf">https://www.cs.cmu.edu/~ckingsf/bioinfo-lectures/gaps.pdf</a></li>
<li>Biological Sequence Analysis (Chapter 2) by Durbin, Eddy, Krogh, and Mitchison. </li>
</ul>
</dd></dl>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
